fish-rust/src/fish.rs

```diff
@@ -0,0 +1,877 @@
+// const OUT_DIR: &str = env!("OUT_DIR", "OUT_DIR was not specified");
+const OUT_DIR: &str = env!("FISH_BUILD_DIR");
+
+/// container to hold the options specified within the command line
+#[derive(Default, Debug)]
+struct FishCmdOpts {
+    /// Future feature flags values string
+    features: WString,
+    /// File path for debug output.
+    debug_output: Option<OsString>,
+    /// File path for profiling output, or empty for none.
+    profile_output: Option<OsString>,
+    profile_startup_output: Option<OsString>,
+    /// Commands to be executed in place of interactive shell.
+    batch_cmds: Vec<OsString>,
+    /// Commands to execute after the shell's config has been read.
+    postconfig_cmds: Vec<OsString>,
+    /// Whether to print rusage-self stats after execution.
+    print_rusage_self: bool,
+    /// Whether no-config is set.
+    no_config: bool,
+    /// Whether no-exec is set.
+    no_exec: bool,
+    /// Whether this is a login shell.
+    is_login: bool,
+    /// Whether this is an interactive session.
+    is_interactive_session: bool,
+    /// Whether to enable private mode.
+    enable_private_mode: bool,
+}
+
+/// \return a timeval converted to milliseconds.
+#[allow(clippy::unnecessary_cast)]
+fn tv_to_msec(tv: &libc::timeval) -> i64 {
+    // milliseconds per second
+    let mut msec = tv.tv_sec as i64 * 1000;
+    // microseconds per millisecond
+    msec += tv.tv_usec as i64 / 1000;
+    msec
+}
+
+fn print_rusage_self(mut dest: impl io::Write) {
+    let mut rs = MaybeUninit::zeroed();
+    if unsafe { libc::getrusage(libc::RUSAGE_SELF, rs.as_mut_ptr()) } != 0 {
+        let s = CString::new("getrusage").unwrap();
+        unsafe { libc::perror(s.as_ptr()) }
+        return;
+    }
+    let rs: libc::rusage = unsafe { rs.assume_init() };
+    let rss_kb = if cfg!(target_vendor = "apple") {
+        // mac use bytes.
+        rs.ru_maxrss / 1024
+    } else {
+        // Everyone else uses KB.
+        rs.ru_maxrss
+    };
+
+    let _ = writeln!(dest, "  rusage self:");
+    let _ = writeln!(dest, "      user time: {} ms", tv_to_msec(&rs.ru_utime));
+    let _ = writeln!(dest, "       sys time: {} ms", tv_to_msec(&rs.ru_stime));
+    let _ = writeln!(
+        dest,
+        "     total time: {} ms",
+        tv_to_msec(&rs.ru_utime) + tv_to_msec(&rs.ru_stime)
+    );
+    let _ = writeln!(dest, "        max rss: {} kb", rss_kb);
+    let _ = writeln!(dest, "        signals: {}", rs.ru_nsignals);
+}
+
+fn determine_config_directory_paths(argv0: &wstr) -> ConfigPaths {
+    // PORTING: why is this not just an associated method on ConfigPaths?
+
+    let mut paths = ConfigPaths::default();
+    let mut done = false;
+    // FIXME: osstr
+    let exec_path = get_executable_path(&argv0.to_string());
+    if let Ok(exec_path) = exec_path.canonicalize() {
+        FLOGF!(
+            config,
+            format!("exec_path: {:?}, argv[0]: {:?}", exec_path, argv0)
```

Comment:
should remove the separate format call -- FLOGF already does that (soon) --
unless we really need the Debug representations? Is Display not enough?

Reply:
>The Cpp code here used std::string with as far as I can understand potentially invalid UTF8.

`std::string` is basically a resizable array/vector of arbitrary bytes, yes.

C++ basically... punts on encoding almost entirely.

Reply:
This does print it with quotes now:
`config: exec_path: "/ABSOLUTE_PATH/fish-shell/build/fish", argv[0]: "./build/fish"`
We probably want to specify when we want to print it in debug vs lossy, but since this is meant for debugging I think Debug is appropriate.

The error/warning-ones uses lossy currently